assignments/patrickdillon/1-1.org

```diff
@@ -0,0 +1,167 @@
+#+TITLE: SICP 1.1 assignment
+#+AUTHOR: Patrick Dillon
+#+BABEL: :cache yes
+#+PROPERTY: header-args :tangle yes
+
+* Chapter Name
+** Section Name
+*** Exercise 1.1
+**** Question
+10
+(+ 5 3 4)
+(- 9 1)
+(/ 6 2)
+(+ (* 2 4) (- 4 6))
+(define a 3)
+(define b (+ a 1))
+(+ a b (* a b))
+(= a b)
+(if (and (> b a) (< b (* a b)))
+    b
+    a)
+(cond ((= a 4) 6)
+      ((= b 4) (+ 6 7 a))
+      (else 25))
+(+ 2 (if (> b a) b a))
+(* (cond ((> a b) a)
+         ((< a b) b)
+         (else -1))
+   (+ a 1))
+**** My Answer
+10
+12
+8
+3
+6
+void
+void
+19
+false (#f)
+4
+16
+6
+16
+
+#+BEGIN_SRC scheme :tangle 1.1.scm
+10
+(+ 5 3 4)
+(- 9 1)
+(/ 6 2)
+(+ (* 2 4) (- 4 6))
+(define a 3)
+(define b (+ a 1))
+(+ a b (* a b))
+(= a b)
+(if (and (> b a) (< b (* a b)))
+    b
+    a)
+(cond ((= a 4) 6)
+      ((= b 4) (+ 6 7 a))
+      (else 25))
+(+ 2 (if (> b a) b a))
+(* (cond ((> a b) a)
+         ((< a b) b)
+         (else -1))
+   (+ a 1))
+#+END_SRC
+
+*** Exercise 1.2
+#+BEGIN_SRC scheme :tangle 1.2.scm
+( / (+ 5 4 (- 2 
+              (- 3 
+                 (+ 6 
+                    (/ 4 5)))))
+    (* 3 
+       (- 6 2) 
+       (- 2 7)))
+#+END_SRC
+
+#+RESULTS:
+: -37/150
+
+*** Exercise 1.3
+#+BEGIN_SRC scheme :tangle 1.3.scm
+(define (sum-larger-squares x y z)
+  (+ (if (> x y) 
+          (* x x)
+          (* y y))
+     (if (> x z)
+          (* x x)
+          (* z z))))
+(sum-larger-squares 1 2 3)
+#+END_SRC
+
+#+RESULTS:
+: 13
+
+*** Exercise 1.4
+The function will add a to the absolute value of b by using addition if b is greater than 0 and subtraction othwerwise.
+
+*** Exercise 1.5
+With applicative-order, operands are only expanded as needed so the procedure would evaluate to return 0. With normal order, the value for p would be evaluated, which would cause a self-referential loop.
+
+*** Exercise 1.6
+The body of Alyssa's procedure is ok, but when arguments are passed to her function they will be evaluated in applicative order. When sqrt-iter is called, the interpreter will see sqrt-iter passed as an argument and try to evaluate its value, thus entering an infinite loop.
+
+*** Exercise 1.7
+The problem with this method when evaluating small numbers is that the hard-coded difference limits the amount of precision you can obtain. For example if you were trying to obtain the square root of 0.000016 and provided a guess of 0.02, the procedure would return without iteratively refining and provide an answer of 0.02.
```

Comment:
nit: a guess of 0.02 should be improved as it's more than 0.001 away from the sqrt of 0.00016. 0.0031 on the other hand is what returns the provided sqrt implementation instead of 0.004.

Reply:
The provided code tests whether the square of the guess is within the threshold from x--not whether the guess is within the threshold of the square root of x (which is what we're trying to determine). The square of 0.02 is 0.00004 which has less than a .001 difference from 0.000016, so I don't think it will be improved. I originally did this by hand but when I double check it now, the code returns 0.02:
`(define (sqrt-iter guess x)
  (if (good-enough? guess x)
      guess
      (sqrt-iter (improve guess x)
                 x)))

(define (improve guess x)
  (average guess (/ x guess)))


(define (average x y)
  (/ (+ x y) 2))


(define (good-enough? guess x)
  (< (abs (- (square guess) x)) 0.001))


(define (sqrt x)
  (sqrt-iter 0.02 x))

(sqrt 0.00016)`

Reply:
oh you are right, i miss understood that sqrt-iter would start with 0.02, and indeed the 0.001 differences is actually checked with the square of the guess. my bad. I got confused when running the (sqrt 0.00016) with an initial guess of 1.0, which resulted in .03293...